tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -783,56 +830,36 @@ def generator_py_func(iterator_id):
         """A `py_func` that will be called to invoke the iterator."""
         # `next()` raises `StopIteration` when there are no more
         # elements remaining to be generated.
-        values = next(generator_state.get_iterator(iterator_id))
+        values = next(generator_state.get_iterator(iterator_id.numpy()))
+
+        def serialize_structure(s):
+          return nest.map_structure(lambda ts: ts._serialize(), s) # pylint: disable=protected-access
 
-        # Use the same _convert function from the py_func() implementation to
-        # convert the returned values to arrays early, so that we can inspect
-        # their values.
         try:
-          flattened_values = nest.flatten_up_to(output_types, values)
+          output_dtypes = nest.map_structure(lambda t: t.dtype, output_spec)
```

Comment:
Could you please change `lambda t: t.dtype` to `lambda t: getattr(t, "dtype", None)`?

In particular, there are some CompositeTensors that don't have a `dtype` (and will never have a single `dtype`), such as `StructuredTensor`.  This should be safe to change, because the `normalize_element` function doesn't ever use elements from `output_dtype` if the corresponding element of `values` is a `CompositeTensor`.

If you'd also like to add a test to `from_generator_test.py` that fails without this change, you could add this:

```
  @combinations.generate(test_base.default_test_combinations())
  def testFromGeneratorStructuredTensor(self):
    if not dataset_ops.tf2.enabled():
      # StructuredTensor doesn't define _to_legacy_output_shapes, which is
      # required by DatasetV1.
      return

    st = structured_tensor.StructuredTensor([], {"x": 12})

    def generator():
      yield st

    dataset = dataset_ops.Dataset.from_generator(
        generator, output_signature=st._type_spec)

    @def_function.function
    def get_first_dataset_item():
      return next(iter(dataset))

    ret = get_first_dataset_item()

    self.assertAllEqual(ret["x"], 12)
```

(For this test, you'll also need to add a couple imports: `tensorflow.python.eager.def_function` and tensorflow.python.ops.structured.structured_tensor`)

Reply:
@edloper sure, we will fix the current problem and I'll try to fix this next.